Mathlib/Algebra/Ring/Divisibility/Lemmas.lean

```diff
@@ -0,0 +1,89 @@
+/-
+Copyright (c) 2023 Oliver Nash. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Oliver Nash
+-/
+import Mathlib.Algebra.Ring.Divisibility.Basic
+import Mathlib.Data.Nat.Choose.Sum
+
+/-!
+# Lemmas about divisibility in rings
+
+# Main results:
+ * `dvd_smul_of_dvd`: stating that `x ∣ y → x ∣ m • y` for any scalar `m`.
+ * `Commute.pow_dvd_add_pow_of_pow_eq_zero_right`: stating that if `y` is nilpotent then
+   `x ^ m ∣ (x + y) ^ p` for sufficiently large `p` (together with many variations for convenience).
+-/
+
+variable {R : Type*} {x y : R}
+
+lemma dvd_smul_of_dvd {M : Type*} [SMul M R] [Semigroup R] [SMulCommClass M R R]
+    {m : M} (h : x ∣ y) : x ∣ m • y :=
+  let ⟨k, hk⟩ := h; ⟨m • k, by rw [mul_smul_comm, ← hk]⟩
+
+lemma dvd_nsmul_of_dvd [NonUnitalSemiring R] {n : ℕ} (h : x ∣ y) : x ∣ n • y := dvd_smul_of_dvd h
+
+lemma dvd_zsmul_of_dvd [NonUnitalRing R] {z : ℤ} (h : x ∣ y) : x ∣ z • y := dvd_smul_of_dvd h
```

Comment:
nit: should `n` / `m` / `z` be explicit arguments here?

extreme nit: should `x` and `y` come after the typeclass arguments, since that seems to normally be the covention?